src/string/is_valid_email_address.rs

```diff
@@ -0,0 +1,139 @@
+    if local_part.len() > MAX_LOCAL_PART_LENGTH as usize
+        || domain.len() > MAX_DOMAIN_LENGTH as usize
+    {
+        return false;
+    }
+
+    // (3)  check quote dots
+    for i in 1..local_part.len() {
+        if local_part.chars().nth(i - 1).unwrap() == '"'
+            && local_part.chars().nth(i).unwrap() == '"'
+        {
+            let proceeding_quote = local_part.chars().nth(i + 1);
+            if proceeding_quote.is_some() && proceeding_quote.unwrap() != '.' {
+                return false;
+            }
+        }
+    }
+
+    // (4)  validate the dots in the local part
+    if local_part.starts_with('.') || local_part.ends_with('.') || local_part.contains("..") {
+        return false;
+    }
+
+    // (5)  check that the characters in the local part are valid
+    for item in local_part.chars() {
+        if !CHARACTERS.contains(item) && !SYMBOLS.contains(item) && item != '.' && item != '"' {
+            return false;
+        }
+    }
+
+    // (6)  check that the domain contains valid characters
+    for item in domain.chars() {
+        if !CHARACTERS.contains(item) && item != '-' && item != '.' && !"[]:".contains(item) {
+            return false;
+        }
+    }
+
+    // (7)  validate the dots in the domain
+    if domain.starts_with('.') || domain.ends_with('.') || domain.contains("..") {
+        return false;
+    }
+
+    // (8)  validate the dashes in the domain
+    if domain.starts_with('-') || domain.ends_with('-') {
+        return false;
+    }
+
+    true
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::string::is_valid_email_address::is_valid_email_address;
+
+    macro_rules! test_is_valid_email_address {
+        ($($name:ident: $inputs:expr,)*) => {
+        $(
+            #[test]
+            fn $name() {
+                let (s, expected) = $inputs;
+                assert_eq!(is_valid_email_address(s), expected);
+            }
+        )*
+        }
+    }
+
+    test_is_valid_email_address! {
+        basic: ("simple@example.com", true),
+        basic_2: ("very.common@example.com", true),
+        cases: ("FirstName.LastName@EasierReading.org", true),
+        one_letter_local: ("x@example.com", true),
+        long_email_subdomains: ("long.email-address-with-hyphens@and.subdomains.example.com", true),
+        tags: ("user.name+tag+sorting@example.com", true),
+        slashes: ("name/surname@example.com", true),
+        no_tld: ("admin@example", true),
+        tld: ("example@s.example", true),
+        quotes_with_space: ("\" \"@example.org", true),
+        quoted_double_dot: ("\"john..doe\"@example.org", true),
+        host_route: ("mailhost!username@example.org", true),
+        quoted_non_letters: (r#""very.(),:;<>[]\".VERY.\"very@\\ \"very\".unusual"@strange.example.com"#, true),
+        percent_symbol: ("user%example.com@example.org", true),
+        local_end_symbol: ("user-@example.org", true),
+        ip_address: ("postmaster@[123.123.123.123]", true),
```

Comment:
```suggestion
        ip_address: ("postmaster@[123.123.123.123]", true),
        wrong_ip_address: ("postmaster@[1234.123.123.123]", false),
```

Reply:
I added checks for ip addresses and added some extra tests to ensure that the ip checks work as expected.